src/avm2/natives/xml.ts

```diff
@@ -71,28 +71,34 @@ module Shumway.AVM2.AS {
   var _asGetEnumerableKeys = Object.prototype.asGetEnumerableKeys;
 
   function isXMLType(val: any): boolean {
-    return (val instanceof ASXML || val instanceof ASXMLList);
+    return (val instanceof AS.ASXML || val instanceof AS.ASXMLList);
   }
 
   // 10.1 ToString
   function toString(node) {
-    if (typeof node === "object" && node !== null) {
-      if (node instanceof ASXMLList) {
-        return node._children.map(toString).join('');
-      }
-      switch (node._kind) {
-        case ASXMLKind.Text:
-        case ASXMLKind.Attribute:
-          return node._value;
-        default:
-          if (node.hasSimpleContent()) {
-            return node._children.map(toString).join('');
-          }
-          return toXMLString(node);
-      }
-    } else {
+    if (!(node instanceof AS.ASXML)) {
       return String(node);
     }
+    switch (node._kind) {
+      case ASXMLKind.Text:
+      case ASXMLKind.Attribute:
+        return node._value;
+      default:
+        if (node.hasSimpleContent()) {
+          var s = '';
+          for (var i = 0; i < node._children.length; i++) {
+            var child = node._children[i];
+            if (child._kind === ASXMLKind.Comment ||
+                child._kind === ASXMLKind.ProcessingInstruction)
+            {
+              continue;
+            }
+            s += toString(child);
```

Comment:
I don't think such a general rule makes much sense. There's a size/speed tradeoff to be made, plus it's not even entirely clear that the push/join pattern is better when Ion is active.

Also, this all is *so* not fast, so even if there's a speed difference to be had here, it just doesn't matter in the grand scheme of things.